### PR TITLE
Example controller for LRAUV

### PR DESCRIPTION
**Pull request by adityapande-1995**
🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

# 🎉 New feature

## Summary
This PR adds a simple PID based controller for LRAUVs, specifically a PI controller for speed, and P for yaw and pitch angles.
The target state is the linear speed, yaw and pitch angle of the vehicle.

## Test it
The required world file is  ``gz-sim/examples/worlds/lrauv_control_demo.sdf``. Run that file in a terminal : 
```
gz sim -r lrauv_control_demo.sdf
```

and in a separate terminal, source the Gazebo Garden workspace and run : 
```
cd ~/garden_ws/src/gz-sim/examples/standalone/lrauv_control/
mkdir build ; cd build
cmake ..; make
./lrauv_control 0.5 0.78 0.174
```

This would make the vehicle maintain a speed of 0.5 m/s, yaw angle of 0.78 rad (~ 45 deg) and pitch angle of 0.147 rad (~10 deg).

## TODO
- [x] Make the LRAUV leave a trail of markers when moving.
- [ ] Interpolate between states for large angles.
- [x] Update README
- [x] Remove sections from codecov and testing

## Screenshot
image
image

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [x] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by arjo129 — DISMISSED** [review]
The README is empty. Did you forget to commit it?

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    {
+      if (_speed < 0.001 &&
+          (_yaw != 0 || _pitch != 0))
+      {
+        std::cout << "Speed needs to be non zero for non zero"
+          " pitch and yaw angles" << std::endl;
+        return;
+      }
+
+      targetSpeed = std::move(_speed);
```
Comment (hidmic) [inline examples/standalone/lrauv_control/lrauv_control.cc:84]: @adityapande-1995 fyi: moving doubles isn't different from copying them.

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      "/model/tethys/joint/horizontal_fins_joint/0/cmd_pos");
+  auto finPubTethys =
+    node.Advertise<gz::msgs::Double>(finTopicTethys);
+
+  while (true)
+  {
+    std::this_thread::sleep_for(std::chrono::milliseconds(200));
+
+    // Update errors in the controller.
+    control.UpdateErrors();
```
Comment (hidmic) [inline examples/standalone/lrauv_control/lrauv_control.cc:188]: @adityapande-1995 meta: if feedback went through a single function (like setpoints do), a separate `UpdateErrors()` call wouldn't be necessary.

File: examples/standalone/lrauv_control/README.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+Open a new terminal window, source your gazebo workspace and run :
+```
+cd ~/gazebo_ws/src/gz-sim/examples/worlds
+gz sim -r lrauv_control_demo.sdf
+```
+
+This should open up a new gazebo window with the LRAUV at rest at the origin.
+
+Step 2 : Open a new terminal source your gazebo workspace, and navigate to this example.
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/README.md]: Fixed https://github.com/gazebosim/gz-sim/pull/1822/commits/d88e69e82a75646b25658287db14962b8f8f7cf4

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    // P for yaw and pitch control.
+    double kYawAngle = -0.5;
+    double kPitchAngle = 0.6;
+
+    // Set the target states to be tracked,
+    // i.e. linear speed (m/s), pitch and yaw angles (rad).
+    void SetTargets(double _speed,
+        double _yaw, double _pitch)
+    {
+      if (_speed < 0.001 &&
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc]: I meant to check for non zero speed, fixed https://github.com/gazebosim/gz-sim/pull/1822/commits/d88e69e82a75646b25658287db14962b8f8f7cf4

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+#include <chrono>
+#include <cmath>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gz/msgs.hh>
+#include <gz/transport.hh>
+
+#define PI 3.14159265359
```
[the remark this answers is not in the document]
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc]: Fixed https://github.com/gazebosim/gz-sim/pull/1822/commits/d88e69e82a75646b25658287db14962b8f8f7cf4

File: examples/standalone/lrauv_control/lrauv_control.cc (hunk 1 above)
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc]: Fixed https://github.com/gazebosim/gz-sim/pull/1822/commits/d88e69e82a75646b25658287db14962b8f8f7cf4

File: examples/standalone/lrauv_control/lrauv_control.cc (hunk 2 above)
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc]: Done https://github.com/gazebosim/gz-sim/pull/1822/commits/7121710a0b9dc3ada25e55e25a457225e6c75b4d

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+      yawAngle = _yaw;
+      pitchAngle = _pitch;
+
+      errorSpeed = targetSpeed - speed;
+      errorSpeedIntegral =
+        std::min(errorSpeedIntegral + errorSpeed, errorSpeedIntegralMax);
+
+      errorYawAngle = targetYawAngle - yawAngle;
+      errorPitchAngle = targetPitchAngle - pitchAngle;
+
```
Comment (hidmic) [inline examples/standalone/lrauv_control/lrauv_control.cc:102]: @adityapande-1995 nit: should we remove the extra blank line?

File: examples/standalone/lrauv_control/lrauv_control.cc [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      speed = _speed;
+      yawAngle = _yaw;
+      pitchAngle = _pitch;
+
+      errorSpeed = targetSpeed - speed;
+      errorSpeedIntegral =
+        std::min(errorSpeedIntegral + errorSpeed, errorSpeedIntegralMax);
+
+      errorYawAngle = targetYawAngle - yawAngle;
+      errorPitchAngle = targetPitchAngle - pitchAngle;
```
Comment (hidmic) [inline examples/standalone/lrauv_control/lrauv_control.cc:101]: @adityapande-1995 meta: should we review how data access is synchronized to make sure it is safe? I see some variables are atomic (but not synchronized as a group) and some are regular variables.

File: examples/standalone/lrauv_control/lrauv_control.cc (hunk 6 above)
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc]: Removed https://github.com/gazebosim/gz-sim/pull/1822/commits/043a82db1cf1a97cb5b8f1d18676010e415c0ece

File: examples/standalone/lrauv_control/lrauv_control.cc (hunk 7 above)
Comment (adityapande-1995) [reply examples/standalone/lrauv_control/lrauv_control.cc:105]: Removed atomics, added a mutex lock around ``UpdateState``, ``SetTarget`` and the main control loop : https://github.com/gazebosim/gz-sim/pull/1822/commits/043a82db1cf1a97cb5b8f1d18676010e415c0ece

**Review by hidmic — APPROVED** [review]

**Comment by hidmic** [comment]
@adityapande-1995 we should check failing CI jobs.

**Comment by adityapande-1995** [comment]
@osrf-jenkins retest this please

**Comment by adityapande-1995** [comment]
@osrf-jenkins retest this please

**Comment by adityapande-1995** [comment]
@osrf-jenkins retest this please

**Comment by adityapande-1995** [comment]
The only test failing here is ``UNIT_SdfGenerator_TEST`` which I don't think is related to the changes made here. What do you think @hidmic @caguero ?

**Comment by hidmic** [comment]
@adityapande-1995 not sure what the policy is, but I see the same as you: it's `UNIT_SdfGenerator_TEST` failing everywhere.

**Pull request merged**